src/sqs/index.ts

```diff
@@ -258,23 +314,65 @@ export class SqsQueue<MessagePayload> {
 			executionId: uuidv4(),
 			queue: this.queueName})
 
-		const context: Context<MessagePayload> = {message, payload, log }
+		const maybeReceivedCount: MessageAttributeValue | undefined = message.MessageAttributes ? message.MessageAttributes["ApproximateReceiveCount"] : undefined;
 
-		log.info("Sqs message received");
+		const receiveCount = maybeReceivedCount ? parseInt(maybeReceivedCount?.StringValue || "1") : 1
+
+		const context: Context<MessagePayload> = {message, payload, log, receiveCount: receiveCount}
+
+		log.info(`Sqs message received. Receive count: ${receiveCount}`);
 
 		try {
 			const messageProcessingStartTime = new Date().getTime();
-			await this.messageHandler.handle(context)
+
+			// Change message visibility timeout to the max processing time
+			// plus EXTRA_VISIBILITY_TIMEOUT_DELAY to have some room for error handling in case of a timeout
+			await this.changeVisabilityTimeout(message, this.timeoutSec + EXTRA_VISIBILITY_TIMEOUT_DELAY, log);
+
+			const timeoutPromise = new Promise((_, reject) =>
+				setTimeout(() => reject(new SqsTimeoutError()), this.timeoutSec*1000)
+			);
+
+			await Promise.race([this.messageHandler(context), timeoutPromise])
+
 			const messageProcessingDuration = new Date().getTime() - messageProcessingStartTime;
 			this.sendProcessedMetrics(messageProcessingDuration);
 			await this.deleteMessage(message, log)
 		} catch (err) {
+
+			const errorHandlingResult = await this.errorHandler(err, context);
+			if(!errorHandlingResult.retryable ||
+				(errorHandlingResult.skipDlq && context.receiveCount >= this.maxAttempts)) {
```

Comment:
Should we log something before we delete?

Reply:
Yea, good point